File: Source/Engine/dotnet/runtime/asset/AssetPath.cs
using System.Runtime.InteropServices;

namespace Hyperion
{
    [ClassBinding(Name = "AssetPath")]
    [StructLayout(LayoutKind.Sequential)]
    public struct AssetPath
    {
        public static readonly AssetPath Invalid = new AssetPath();

        private Name assetName;
        private uint bucketIndex;

        public AssetPath()
        {
            assetName = new Name();
            bucketIndex = 0;
        }

        public bool Valid => assetName.Valid && bucketIndex != 0;

        public override string ToString()
        {
            /// Same impl as native AssetPath::ToString().
            
            if (!Valid)
                return "<Invalid>";

            string bucketName = AssetBucket.GetAssetBucketName(bucketIndex);
            return $"{bucketName}/{assetName}";
        }
    }
}

File: Source/Engine/asset/AssetPath.cpp
/*!
 *  @author: The Hyperion Contributors
 *  @date 2016-2026
 *  @licence MIT
*/

#include <AssetPch.hpp>

#include <asset/AssetPath.hpp>

#include <AssetPath.generated.inl>

namespace Hyperion {

HYP_DECLARE_LOG_CHANNEL(Assets);

AssetPath::AssetPath(const UTF8StringView& path)
{
    if (!path)
    {
        return;
    }

    const size_t slashIdx = path.FindFirstIndex('/');

    if (slashIdx != UTF8StringView::NotFound)
    {
        UTF8StringView bucketName = path.Substr(0, slashIdx);

        const AssetBucket& bucket = GetAssetBucketByName(StringHash(bucketName));
        AssertDebug(bucket != AssetBuckets::None, "Invalid asset bucket {}", bucketName);

        bucketIndex = bucket.GetIndex();
    }

    assetName = CreateNameFromDynamicString(ANSIString(path.Substr(slashIdx != UTF8StringView::NotFound ? slashIdx + 1 : 0, SIZE_MAX)));
}

String AssetPath::ToString() const
{
    if (!IsValid())
    {
        return "<invalid asset path>";
    }

    const char* bucketName = GetAssetBucketName(bucketIndex);
    AssertDebug(bucketName != nullptr, "Invalid bucket index {}", bucketIndex);

    String str;
    str.Reserve(32);

    str += bucketName;
    str += '/';
    str += *assetName;

    return str;
}

} // namespace Hyperion
